### ISSUE TITLE
ViewContext.ExecutingFilePath is incorrect for custom tags(taghelper) in sections

### ISSUE THREAD

**Issue by expressiveco** [aspnet/Mvc#7619]
When custom tags are used in sections of views then ViewContext.ExecutingFilePath is reported incorrectly within the taghelper. It is correct when used outside of sections.

# Repro Steps
* In index.cshtml
``` asp
@section scripts {
    <my-scripts Path="@Path"></my-scripts>
}
```

# Expected
* In "my-scripts" taghelpers Process() method, ViewContext.ExecutingFilePath is expected to be **"Index.cshtml"** but it is **"_Layout.cshtml"**.

# Actual
* In "my-scripts" taghelpers Process() method, ViewContext.ExecutingFilePath is expected to be **"Index.cshtml"**.

# Currently Proposed Solutions
* It could be set in sections;
``` asp
 @section scripts {
    @{ ViewContext.ExecutingFilePath = Path; }
    <my-scripts Path="@Path"></my-scripts>
}
```
* A "Path" attribute for the tag could be used to get current views Path;
``` asp
@section scripts {
    <my-scripts Path="@Path"></my-scripts>
}
```

# Solution for fixing the bug
* [SectionTargetExtension.WriteSection](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.razor.language.extensions.sectiontargetextension.writesection?view=aspnetcore-2.0) could generate following assignment while generating the lambda parameter for DefineSection() call;
``` asp
    @{ ViewContext.ExecutingFilePath = Path; }
```

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @expressiveco.
@NTaylorMullen, can you please look into this? Thanks!

**Comment by expressiveco** [comment]
But what about RenderBody()? It is called from layout to render the view but in a view ExecutingFilePath value is not the layouts path but the views.

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen, but what about RenderBody()? It is called from layout to render the view but in a view ExecutingFilePath value is not the layouts path but the views.

You're right; however, the View itself will always render so we render that view first with its own ViewContext and cache it, then we render the `_Layout.cshtml` and when it calls `RenderBody` we paste in the already rendered View. It's a little quirky hence my previous posts wording but sections differ from Views where they don't always run.

**Comment by expressiveco** [comment]
So, i want to have a general purpose solution to get file path of the view that contains the piece of code i.e. ContainingFilePath. 
Currently, following is my preference with a mandatory Path attribute
``` asp
@section scripts {
    <my-scripts Path="@Path"></my-scripts>
}
```
